src/librustc_typeck/astconv.rs

```diff
@@ -191,6 +192,153 @@ impl<'o, 'gcx: 'tcx, 'tcx> dyn AstConv<'gcx, 'tcx>+'o {
+            if has_self {
+                if let Some(param) = next_param {
+                    if param.index == 0 {
+                        if let GenericParamDefKind::Type { .. } = param.kind {
+                            push_kind(&mut substs, self_ty.map(|ty| ty.into())
+                                .unwrap_or_else(|| inferred_kind(None, param, true)));
+                            next_param = params.next();
+                        }
+                    }
+                }
+            }
+
+            // Check whether this segment takes generic arguments and the user has provided any.
+            let (generic_args, infer_types) = args_for_def_id(def_id);
+            if let Some(ref generic_args) = generic_args {
+                // We're going to iterate through the generic arguments that the user
+                // provided, matching them with the generic parameters we expect.
+                // Mismatches can occur as a result of elided lifetimes, or for malformed
+                // input. We try to handle both sensibly.
+                'args: for arg in &generic_args.args {
+                    while let Some(param) = next_param {
+                        match param.kind {
+                            GenericParamDefKind::Lifetime => match arg {
+                                GenericArg::Lifetime(_) => {
+                                    push_kind(&mut substs, provided_kind(param, arg));
+                                    next_param = params.next();
+                                    continue 'args;
+                                }
+                                GenericArg::Type(_) => {
+                                    // We expected a lifetime argument, but got a type
+                                    // argument. That means we're inferring the lifetimes.
+                                    push_kind(&mut substs, inferred_kind(None, param, infer_types));
+                                    next_param = params.next();
+                                }
+                            }
+                            GenericParamDefKind::Type { .. } => match arg {
+                                GenericArg::Type(_) => {
+                                    push_kind(&mut substs, provided_kind(param, arg));
+                                    next_param = params.next();
+                                    continue 'args;
+                                }
+                                GenericArg::Lifetime(_) => {
+                                    // We expected a type argument, but got a lifetime
+                                    // argument. This is an error, but we need to handle it
+                                    // gracefully so we can report sensible errors. In this
+                                    // case, we're simply going to infer the remaining
+                                    // arguments.
+                                    if err_if_invalid {
```

Comment:
Alright so I think that if we *do* want to move error reporting here, this is one of the prime places that should be changed.